pkg/dequeuer/async_handler.go

```diff
@@ -0,0 +1,206 @@
+	"github.com/cortexlabs/cortex/pkg/types/status"
+	"go.uber.org/zap"
+)
+
+const (
+	// CortexRequestIDHeader is the header containing the workload request id for the user container
+	CortexRequestIDHeader = "X-Cortex-RequestID"
+)
+
+type AsyncMessageHandler struct {
+	config      AsyncMessageHandlerConfig
+	aws         *awslib.Client
+	metrics     statsd.ClientInterface
+	log         *zap.SugaredLogger
+	storagePath string
+}
+
+type AsyncMessageHandlerConfig struct {
+	ClusterUID string
+	Bucket     string
+	APIName    string
+	TargetURL  string
+}
+
+type userPayload struct {
+	Body        io.ReadCloser
+	ContentType string
+}
+
+func NewAsyncMessageHandler(config AsyncMessageHandlerConfig, awsClient *awslib.Client, metrics statsd.ClientInterface, logger *zap.SugaredLogger) *AsyncMessageHandler {
+	return &AsyncMessageHandler{
+		config:      config,
+		aws:         awsClient,
+		metrics:     metrics,
+		log:         logger,
+		storagePath: awslib.S3Path(config.Bucket, fmt.Sprintf("%s/workloads/%s", config.ClusterUID, config.APIName)),
+	}
+}
+
+func (h *AsyncMessageHandler) Handle(message *sqs.Message) error {
+	err := h.handleMessage(message)
+	if err != nil {
+		var requestID string
+		if message.Body != nil {
+			requestID = *message.Body
+		}
+
+		h.log.Errorw("failed processing request", "id", requestID, "error", err)
+		err = h.handleFailure(requestID)
+		if err != nil {
+			h.log.Error(zap.Error(err))
+		}
+	}
+	return nil
+}
+
+func (h *AsyncMessageHandler) handleMessage(message *sqs.Message) error {
+	requestID := *message.Body
+	h.log.Infow("processing workload", "id", requestID)
+
+	err := h.updateStatus(requestID, status.AsyncStatusInProgress)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to update status to %s", status.AsyncStatusInProgress))
+	}
+
+	payload, err := h.getPayload(requestID)
+	if err != nil {
+		return errors.Wrap(err, "failed to get payload")
+	}
+	defer h.deletePayload(requestID)
+
+	result, err := h.submitRequest(payload, requestID)
+	if err != nil {
+		return errors.Wrap(err, "failed to submit request to user container")
+	}
+
+	if err = h.uploadResult(requestID, result); err != nil {
+		return errors.Wrap(err, "failed to upload result to storage")
+	}
+
+	if err = h.updateStatus(requestID, status.AsyncStatusCompleted); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to update status to %s", status.AsyncStatusCompleted))
+	}
+
+	h.log.Infow("workload processing complete", "id", requestID)
+
+	return nil
+}
+
+func (h *AsyncMessageHandler) handleFailure(requestID string) error {
+	err := h.updateStatus(requestID, status.AsyncStatusFailed)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to update status to %s", status.AsyncStatusFailed))
+	}
+	return nil
+}
+
+func (h *AsyncMessageHandler) updateStatus(requestID string, status status.AsyncStatus) error {
+	key := fmt.Sprintf("%s/%s/status/%s", h.storagePath, requestID, status)
```

Comment:
These `fmt.Sprintf("...")` keys can be converted to functions and moved to a new file. That way, there's only a single obvious place where the paths are defined. What do you think?